CaRpools
========

sudo apt-get update
sudo apt-get install default-jre
sudo apt-get install default-jdk
sudo apt-get install libxml2-dev

conda install -c bioconda mageck

# sudo apt-get install r-cran-rjava

sudo R CMD javareconf


in R

install.packages("rJava")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
    
BiocManager::install("DESeq2")
BiocManager::install("biomaRt")



https://cran.r-project.org/web/packages/caRpools/vignettes/CaRpools.html


```{r}
CONTROL <- load.file("./data/Day14_R1.txt", header= TRUE, sep="\t")
TREAT <- load.file("./data/Day14_R2.txt", header= TRUE, sep="\t")

libFILE = load.file("./data/pilotscreen_Fitness.fasta",header = FALSE, type="fastalib")

CONTROL.g <- aggregatetogenes(data.frame=CONTROL, agg.function=sum, extractpattern=expression("^(.+?)(_.+)"), type="aggregate")
TREAT.g <- aggregatetogenes(data.frame=TREAT, agg.function=sum, extractpattern=expression("^(.+?)(_.+)"), type="aggregate")

CONTROL.stats <- stats.data(dataset=CONTROL, namecolumn = 1, fullmatchcolumn = 2, extractpattern=expression("^(.+?)_.+"), type="stats")
TREAT.stats <- stats.data(dataset=TREAT, namecolumn = 1, fullmatchcolumn = 2, extractpattern=expression("^(.+?)_.+"), type="stats")

carpools.read.distribution(CONTROL, fullmatchcolumn=2, breaks=200, title="CONTROL", xlab="log2 Readcount", ylab="# sgRNAs",statistics=TRUE)
carpools.read.distribution(CONTROL, fullmatchcolumn=2, breaks=200, title="CONTROL", xlab="log2 Readcount", ylab="# sgRNAs",statistics=TRUE, type="whisker")

carpools.read.distribution(TREAT, fullmatchcolumn=2, breaks=200, title="TREAT", xlab="log2 Readcount", ylab="# sgRNAs",statistics=TRUE)
carpools.read.distribution(TREAT, fullmatchcolumn=2, breaks=200, title="TREAT", xlab="log2 Readcount", ylab="# sgRNAs",statistics=TRUE, type="whisker")

CONTROL.readspergene <- carpools.reads.genedesigns(CONTROL, namecolumn=1, fullmatchcolumn=2, title=paste("% sgRNAs:", "CONTROL", sep=" "), xlab="% of sgRNAs present", ylab="# of Genes")
TREAT.readspergene <- carpools.reads.genedesigns(CONTROL, namecolumn=1, fullmatchcolumn=2, title=paste("% sgRNAs:", "TREAT", sep=" "), xlab="% of sgRNAs present", ylab="# of Genes")

carpools.read.count.vs(dataset=list(TREAT,CONTROL), dataset.names = c("TREAT", "CONTROL"), pairs=FALSE, namecolumn=1, fullmatchcolumn=2, title="", pch=16, normalize=TRUE, norm.function=median, labelgenes="random", labelcolor="blue", center=FALSE, aggregated=FALSE)

carpools.read.count.vs(dataset=list(TREAT,CONTROL), dataset.names = c("TREAT", "CONTROL"), pairs=FALSE, namecolumn=1, fullmatchcolumn=2, title="", pch=16, normalize=TRUE, norm.function=median, labelgenes="ENSG00000007237", labelcolor="red", center=FALSE, aggregated=FALSE)

carpools.raw.genes(untreated.list = list(CONTROL), treated.list = list(TREAT), genes="ENSG00000007237", namecolumn=1, fullmatchcolumn=2, norm.function=median, extractpattern=expression("^(.+?)_.+"), do.plot=TRUE, log=FALSE, put.names=TRUE, type="foldchange" )

data.wilcox = stat.wilcox(untreated.list = list(CONTROL), treated.list = list(TREAT), namecolumn=1, fullmatchcolumn=2, normalize=TRUE, norm.fun=median, sorting=FALSE, controls="random", control.picks=NULL)

data.deseq = stat.DESeq(untreated.list = list(CONTROL), treated.list = list(TREAT), namecolumn=1, fullmatchcolumn=2, extractpattern=expression("^(.+?)(_.+)"), sorting=FALSE, filename.deseq = "ANALYSIS-DESeq2-sgRNA.tab", fitType="parametric")

data.mageck = stat.mageck(untreated.list = list(CONTROL), treated.list = list(TREAT), namecolumn=1, fullmatchcolumn=2, norm.fun="median", extractpattern=expression("^(.+?)(_.+)"), mageckfolder=NULL, sort.criteria="neg", adjust.method="fdr", filename = "TEST" , fdr.pval = 0.05)

```

