CRISPhieRmix
============

The analysis tools (e.g., [caRpools](https://cran.r-project.org/web/packages/caRpools/vignettes/CaRpools.html)) used for CRISPR KO screening may not be optimal to CRISPR a/i screening, due to two reasons:

* variability in guide efficiency
* large rare off-target effects

[CRISPhieRmix](https://github.com/timydaley/CRISPhieRmix) [(*Genome Biology 19:159 (2018)*)](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1538-6) use a hierarchical mixture model for large-scale CRISPR a/i pooled screens, which resolves the above mentioned issues and allows for more accurate and powerful inferences.

## Installation

```{r}
install.packages("devtools")
remove.packages('curl')
```

In terminal, type:

```
sudo apt-get remove libcurl4-nss-dev
sudo apt-get install libcurl4-openssl-dev
```

```{r}
install.packages('curl')
devtools::install_github("lintian0616/CRISPhieRmix")
```

## Data Download

We will use [PRJNA553254](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA553254) ([eLife 8:e44288](https://elifesciences.org/articles/44288)) as our example, which use [human CRISPRi v2 library]().

|   Run ID   |sample name|
|------------|-----------|
| SRR9936564 |  CDK9i_1  |
| SRR9936565 |  CDK9i_2  |
| SRR9936566 |  MCL1i_1  |
| SRR9936567 |  MCL1i_2  |
| SRR9936568 |   DMSO_1  |
| SRR9936569 |   DMSO_2  |

We will first download the fastq raw reads.

```
# conda install sra-tools # if we do not have sra-tools
fastq-dump -I --split-3 SRR9936564
fastq-dump -I --split-3 SRR9936565
fastq-dump -I --split-3 SRR9936566
fastq-dump -I --split-3 SRR9936567
fastq-dump -I --split-3 SRR9936568
fastq-dump -I --split-3 SRR9936569
```

## Count Spacers

```
python ~/CRISPR_Screening/count_spacers.py -f <(zcat CDK9_1.fastq.gz) -o CDK9_1.csv -i ~/CRISPR_Screening/count_spacers_files/Genome-wide_CRISPRi-v2-human-spacer.csv

python ~/CRISPR_Screening/count_spacers.py -f <(zcat CDK9_2.fastq.gz) -o CDK9_2.csv -i ~/CRISPR_Screening/count_spacers_files/Genome-wide_CRISPRi-v2-human-spacer.csv

python ~/CRISPR_Screening/count_spacers.py -f <(zcat MCL_1.fastq.gz) -o MCL_1.csv -i ~/CRISPR_Screening/count_spacers_files/Genome-wide_CRISPRi-v2-human-spacer.csv

python ~/CRISPR_Screening/count_spacers.py -f <(zcat MCL_2.fastq.gz) -o MCL_2.csv -i ~/CRISPR_Screening/count_spacers_files/Genome-wide_CRISPRi-v2-human-spacer.csv

python ~/CRISPR_Screening/count_spacers.py -f <(zcat DMSO_1.fastq.gz) -o DMSO_1.csv -i ~/CRISPR_Screening/count_spacers_files/Genome-wide_CRISPRi-v2-human-spacer.csv

python ~/CRISPR_Screening/count_spacers.py -f <(zcat DMSO_2.fastq.gz) -o DMSO_2.csv -i ~/CRISPR_Screening/count_spacers_files/Genome-wide_CRISPRi-v2-human-spacer.csv
```